nomenclature/processor/requiredData.py

```diff
@@ -0,0 +1,140 @@
+
+        # check for undefined regions and variables
+        for dim in ("region", "variable"):
+            if not_defined_dims := self._undefined_dimension(dim, dsd):
+                error_msg += (
+                    f"The following {dim}(s) were not found in the "
+                    f"DataStructureDefinition:\n{not_defined_dims}\n"
+                )
+        # check for defined variables with wrong units
+        if wrong_unit_variables := self._wrong_unit_variables(dsd):
+            error_msg += (
+                "The following variables were found in the "
+                "DataStructureDefinition but have the wrong unit "
+                "(affected variable, wrong unit, expected unit):\n"
+                f"{wrong_unit_variables}"
+            )
+
+        if error_msg:
+            raise ValueError(error_msg)
+
+    def _undefined_dimension(
+        self, dimension: str, dsd: DataStructureDefinition
+    ) -> List[str]:
+        missing_items: List[str] = []
+        # only check if both the current instance and the DataStructureDefinition
+        # have the dimension in question
+        if getattr(self, dimension) and hasattr(dsd, dimension):
+            missing_items.extend(
+                dim
+                for dim in getattr(self, dimension)
+                if dim not in getattr(dsd, dimension)
+            )
+
+        return missing_items
+
+    def _wrong_unit_variables(
+        self, dsd: DataStructureDefinition
+    ) -> List[Tuple[str, str, str]]:
+        wrong_units: List[Tuple[str, str, str]] = []
+        if hasattr(dsd, "variable"):
+            wrong_units.extend(
+                (var, self.unit, getattr(dsd, "variable")[var].unit)
+                for var in getattr(self, "variable")
+                if self.unit and self.unit not in getattr(dsd, "variable")[var].units
+            )
+
+        return wrong_units
+
+
+class RequiredDataValidator(BaseModel):
+
+    name: str
+    required_data: List[RequiredData]
+    optional_data: Optional[List[RequiredData]]
+    file: Path
+
+    @classmethod
+    def from_file(cls, file: Union[Path, str]) -> "RequiredDataValidator":
+        with open(file, "r") as f:
+            content = yaml.safe_load(f)
+        return cls(file=file, **content)
+
+    def apply(self, df: IamDataFrame) -> IamDataFrame:
+        error = False
+        # check for required data and raise error if missing
+        for data in self.required_data:
+            if (missing_index := df.require_data(**data.dict())) is not None:
+                error = True
+                logger.error(
+                    f"Required data {data} from file {get_relative_path(self.file)} "
+                    f"missing for:\n{missing_index}"
+                )
+        if error:
+            raise RequiredDataMissingError(
+                "Required data missing. Please check the log for details."
+            )
+
+        # check for optional data and issue warning if missing
```

Comment:
I think it's confusing to have "optional data" within a validation for required data - and why print a warning if optional data isn't there?

Reply:
Fair point, the reason I put it there is that I was thinking along the lines of a use case where we have a core set of data for a given post-processing plus an additional set of optional data. Having these optional data is desired as it makes the post-processing more accurate and requires less infilling, like the ar6 climate assessment workflow does it. 
Now that I think about it again it is probably best though to cover this logic in a `MAGICCPostProcessor` class and leave `requiredData` more simple.